---
title: "RNA Sequencing Analysis"
author: "Alex Huber"
output:
  html_document:
    toc: true
    toc_float: true
knit:
  (function(input, ...) {
    rmarkdown::render(
      input,
      output_file = 'index.html'
    )
  })

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, cache = TRUE)
```

<br><br>

## Background  

#### This is an analysis of the RNA Sequencing data first published in the paper [Nickel exposure induces persistent mesenchymal phenotype in human lung epithelial cells through epigenetic activation of ZEB1](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5930076/). Here I perform Differential Expression Sequencing Analysis, looking at patterns of up- and down-regulated genes and biological pathways in each experimental condition.

```{r load_pkgs, include=FALSE}
library(pacman)

pacman::p_load(rhdf5, edgeR, tidyverse, stringr, DESeq2, ggpubr, pheatmap, RColorBrewer, biomaRt, ComplexHeatmap, clusterProfiler, ReactomePA, enrichplot, cowplot, gridExtra)

require(scales)

```

<br><br>

## Fetching Data & Creating Study Design  

#### The first step is collecting the public data through the [ARCHS4 online database](https://maayanlab.cloud/archs4/). I then identify the experimental conditions for each sample and put them into a variable my.study.design.

#### The different conditions tested here are
* Control_3D: cells untreated with Nickel after 3 days
* Control_6W: cells untreated with Nickel after 6 weeks
* Ni_3D: cells treated with Nickel after 3 days
* Ni_3D+14D: cells treated with Nickel after 17 days
* Ni_6W: cells treated with Nickel after 6 weeks
Ni_6W+2W: cells treated with Nickel for 6 weeks then without Nickel for 2 weeks
* Ni_clone: cells cloned from a single cell in the Ni_6W condition

```{r query_data_create_study_design}

# Create variable for archs4 hdf5 local file location (downloaded from https://maayanlab.cloud/archs4/)
archs4.human <- "../human_gene_v2.3.h5"

# h5ls.layout <- h5ls(archs4.human)

# Create reference for every gene
genes <- h5read(archs4.human, "meta/genes/ensembl_gene")

# Create df for every sample's series ID
all.series.ids <- h5read(archs4.human, name="meta/samples/series_id")

# IDing my samples by their series IDs
my.sample.locations <- which(str_detect(all.series.ids, "GSE95180"))


# Creating df for my samples' expression
my.sample.expressions <- t(h5read(archs4.human, "data/expression",
                     index=list(my.sample.locations, 1:length(genes))))

# Creating vector for my samples' GEO accession IDs
my.sample.ids <- h5read(archs4.human, name="meta/samples/geo_accession")[my.sample.locations]

# Renaming rows cols
rownames(my.sample.expressions) <- genes
colnames(my.sample.expressions) <- my.sample.ids

# Creating vector for my samples' descriptions
my.sample.descriptions <- h5read(archs4.human, name="meta/samples/title")[my.sample.locations]

# Creating a study design file from sample descriptions in hdf5 metadata
my.study.design <- tibble(id = my.sample.ids,
                          condition = as.factor(str_extract(my.sample.descriptions, ".+(?=_)")),
                          rep = as.factor(str_extract(my.sample.descriptions, ".$")))

my.study.design
```


<br><br>

## Plotting Total Reads

#### I first look at the total number of sequences that came in from each sample in order look for any obvious outliers. These samples all appear to be within a similar range.
```{r plot_reads}
# Creating a dataframe with the total counts of each sample 
fullCounts <- as.data.frame(colSums(my.sample.expressions))
colnames(fullCounts) <- "counts"
fullCounts$condition <- my.study.design$condition


# Plotting total counts for each condition 
fullCounts %>%
  ggplot(aes(x = condition, y = counts, size = 6)) +
  geom_point(position = position_jitterdodge(jitter.width = 0.05)) +
  guides(size = "none") +
  labs(title = "Raw Counts in Every Condition") +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.9, hjust=1))

```

<br><br>

## Creating DESeq Object & Plotting Gene of Interest Reads

#### This paper's authors found an increase in expression of the gene ZEB1 after 6 weeks of Nickel exposure. I found that there is an increase of ZEB1 RNA transcripts after Nickel exposure, but the statistical significance is only p < 0.05 when comparing the the Ni_6W samples to the Control_3D samples rather than the Control_6W samples (p=0.1) using a t-test.
```{r create_DESeq_obj}

# Filter genes out that no reads in 75% of samples
keep <- rowSums(my.sample.expressions >= 1) >= ncol(my.sample.expressions)*0.75
my.sample.expressions <- my.sample.expressions[keep,]

# Create a DESeqDataSet from the my.sample.expressions matrix
# Using my.sytudy.design to inform the analysis of sample information
# Using design argument to inform of potentially impactful parameters
dds <- DESeqDataSetFromMatrix(my.sample.expressions, colData = my.study.design, design = ~ rep + condition)
dds <- DESeq(dds)

# Performing count analysis on ZEB-1 a gene which was found to be upregulated in the source paper
zeb1.count.data <- plotCounts(dds, gene = "ENSG00000148516", main = "ZEB1", returnData = T)

# Plotting ZEB-1 counts in all samples
ggplot(zeb1.count.data, aes(x = condition, y = count)) +
  geom_boxplot() +
  geom_point() +
  labs(title = "ZEB1 Counts in All Samples") +
  stat_compare_means(comparisons = list(c(2,5)), method = "t.test", label.y = 450) +
    stat_compare_means(comparisons = list(c(1,5)), method = "t.test", label.y = 500)


```

<br><br>

## Performing PCA on All Samples

#### My next step was to perform pricipal component analysis (PCA) in order to determine which sample groups were most distinct from the others. I found that the greatest variance between all samples (PC1) separated the Control_3D samples from the remaining samples. Interestingly, these Control_3D samples also separated to form two distinct clusters along the second largest principal component axis (PC2). I then plotted a distance matrix which clustered the samples by which other samples they are most similar to. The darker a cell is colored represents a greater similarity to the corresponding sample. This shows that the Control_3D samples are very dissimilar to the other samples. Along with, interestingly, one Ni_6W sample.
```{r PCA}

# Performing variance stabilizing transformation in order to normalize counts across the experiment
vsd <- vst(dds)

# Plotting principal component analysis
plotPCA(vsd, intgroup = c("condition")) +
  geom_label(aes(label = dds$condition)) +
  labs(title = "PC1 vs PC2")

# Calculating and plotting sample distances by hierarchical clustering
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- vsd$condition
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors,
         main = "Distance Matrix") #plot the PCA
```
<br><br>

## Performing PCA on Samples of Interest

#### My next step was to narrow down the samples that I was investigating. The comparisons I was most interested in making were between the Control_6W, Ni_6W, and Ni_6W+2W groups. The largest PC showed a stark difference between one Ni_6W outlier and the remaining samples (which was also revealed in the previous distance matrix). PC2, however, showed a clear difference between the Control_6W samples and the Ni-treated samples. The distance matrix confirmed these findings.
```{r PCA_select_samples}

# Creating a subset of samples in order to more clearly define differences between samples of interest
select.sample.study.design <- filter(my.study.design, condition %in% c("Control_6W", "Ni_6W", "Ni_6W+2W"))
select.sample.expressions <- my.sample.expressions[,select.sample.study.design$id]

# Creating DDS object for sample subset
select.dds <- DESeqDataSetFromMatrix(select.sample.expressions, colData = select.sample.study.design, design = ~ rep + condition)
select.dds <- DESeq(select.dds)

# Performing variance stabilizing transformation in order to normalize counts across the experiment
select.vsd <- vst(select.dds)

# Plotting principal component analysis
plotPCA(select.vsd, intgroup = c("condition")) +
  geom_label(aes(label = select.dds$condition)) +
  labs(title = "PC1 vs PC2")

# Calculating and plotting sample distances by hierarchical clustering
select.sampleDists <- dist(t(assay(select.vsd)))
select.sampleDistMatrix <- as.matrix(select.sampleDists)
rownames(select.sampleDistMatrix) <- select.vsd$condition
colnames(select.sampleDistMatrix) <- NULL
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)
pheatmap(select.sampleDistMatrix,
         clustering_distance_rows=select.sampleDists,
         clustering_distance_cols=select.sampleDists,
         col=colors,
         main = "Distance Matrix")
```

<br><br>

## Creating Volcano Plot

#### The next step was to identify the genes that were significantly differrentially expressed between the Control_6W samples and the Ni_6W samples. I visualized these genes using a volcano plot with the log of each gene's fold change on the x-axis and the log of each gene's p-value on the y-axis. My genes of interest are labeled here in red with abs(log2FC) > 1 and p.adj value < 0.05.
```{r creating_volcanoplot}

# Creating comparison objects which detail expression differences between chosen conditions
Ni.6W.vs.Ctrl.6W <- as.data.frame(results(dds, contrast = c("condition", "Control_6W", "Ni_6W")))

# Assigning padj value of 1 to all genes for which no padj value could be calculated
Ni.6W.vs.Ctrl.6W$padj[is.na(Ni.6W.vs.Ctrl.6W$padj)] <- 1

# Assigning specific values to each gene to signify if it has l2fc >=1, padj <= 0.05, or both
Ni.6W.vs.Ctrl.6W <- Ni.6W.vs.Ctrl.6W %>%
  mutate(volcano_color = (abs(log2FoldChange) >= 1) * 2 + (padj <= 0.05))

# Plotting volcanoes
ggplot(Ni.6W.vs.Ctrl.6W, aes(x = log2FoldChange, y = -log10(padj), color = as.factor(volcano_color))) +
  geom_point() +
  labs(title = "Differentially Expressed Genes in Ni-Treated BEAS-2B Cells",
       subtitle = "6 Weeks Ni Medium Compared to 6 Weeks in Ni-Free Medium") +
  xlab("log2FoldChange") + ylab("-log10(padj)") +
  scale_color_manual(labels = (c("abs(l2FC) < 1 & padj > 0.05",
                                 "abs(l2FC) < 1 & padj < 0.05",
                                 "abs(l2FC) > 1 & padj > 0.05",
                                 "abs(l2FC) > 1 & padj < 0.05")),
                     values = c("black", "darkgreen", "blue", "red"))
```
<br><br>

## Creating Heatmap of Genes of Interest

#### I then visualized my genes of interest using a heatmap, showing the relative expression of each gene within each sample group. Each row represents a gene of interest and each column represents one sample. I chose to subset this heatmap into two groups in order to perform gene set enrichment analysis on each group, which will allow me to determine if the genes that were largely up- or down-regulated are involved in any of the same biological pathways or processes.

```{r creating_heatmap}

# Creating comparison objects which detail expression differences between chosen conditions
Ni.6W2W.vs.Ctrl.6W <- as.data.frame(results(dds, contrast = c("condition", "Control_6W", "Ni_6W+2W")))

# Creating a vector that contains all significantly up or down-regulated genes
Ni.6W.vs.Ctrl.6W.sig.genes <- rownames(filter(Ni.6W.vs.Ctrl.6W, abs(log2FoldChange) >= 1, padj <= 0.05))
Ni.6W2W.vs.Ctrl.6W.sig.genes <- rownames(filter(Ni.6W2W.vs.Ctrl.6W, abs(log2FoldChange) >= 1, padj <= 0.05))
all.sig.genes <- unique(Ni.6W.vs.Ctrl.6W.sig.genes, Ni.6W2W.vs.Ctrl.6W.sig.genes)

# Filtering DESeq results object to only include genes of interest and samples of interest and performing variance stabilizing transformation
trim_vsd <- as.data.frame(assay(vsd))[all.sig.genes, filter(my.study.design, condition %in% c("Control_6W", "Ni_6W", "Ni_6W+2W"))$id]

# Scaling the results object so that gene expression means are centered at zero
scale_vsd <- as.data.frame(t(scale(t(trim_vsd))))

# Creating a heatmap with 2 row splits
Heatmap.2splits <- Heatmap(scale_vsd,
        name = "Expression",
        column_title = "Differentially Expressed Genes in Ni-Treated BEAS-2B Cells",
        show_row_names = F,
        show_column_names = F,
        cluster_columns = F,
        column_split = filter(my.study.design, condition %in% c("Control_6W", "Ni_6W", "Ni_6W+2W"))$condition,
        row_split = 2,
        clustering_method_rows = "ward.D2",
        top_annotation = HeatmapAnnotation(foo = anno_block(labels = c("Control_6W", "Ni_6W", "Ni_6W+2W")))
        )

# Plotting heatmap
Heatmap.2splits

```

<br><br>

## Performing Gene Set Enrichment Analysis

#### The final step involves analyzing the subsets of genes that were differentially expressed in order to discover any biological pathways or processes that are overrepresented. This offers insight into the specific mechanisms that the experimental conditions affect most broadly. 
```{r enrichment_analysis, fig.width=20,fig.height=12}
# A for loop that creates a list of enrichment pathways for each group in an input heatmap
enriched <- list()
dots <- list()
cnets <- list()

for (i in 1:length(row_order(Heatmap.2splits))) {
  enriched[[i]] <-
    enrichGO(
      Heatmap.2splits@row_names_param$labels[row_order(Heatmap.2splits)[[i]]],
      OrgDb = "org.Hs.eg.db",
      keyType = "ENSEMBL",
      ont = "BP",
      pvalueCutoff = 1,
      qvalueCutoff = 1
    )
  dots[[i]] <- enrichplot::dotplot(enriched[[i]], x = "Count", showCategory = 8, font.size = 10)
  cnets[[i]] <- enrichplot::cnetplot(enriched[[i]], node_label = "category", font.size = 10)

}

plot_grid(dots[[1]], dots[[2]], cnets[[1]], cnets[[2]],
          nrow = 2, ncol = 2, align = "hv",
          labels = c("Dotplot Figure of Enriched Gene Sets in Cluster 1",
                     "Dotplot Figure of Enriched Gene Sets in Cluster 2",
                     "Netplot Figure of Enriched Gene Sets in Cluster 1",
                     "Netplot Figure of Enriched Gene Sets in Cluster 2"))

```


